miri/datamodels/cdplib.py

```diff
@@ -927,8 +929,17 @@ def match_cdp_filename(self, cdptype, model='FM', detector=None,
         # The readout pattern is optional, and a CDP valid for any
         # pattern is specified by missing it out completely.
         if (readpatt is not None) and (readpatt != 'ANY') and (readpatt != 'N/A'):
-            match_strings.append(readpatt)
-               
+            # NOTE: There are several variations on the FAST readout pattern
+            # (FAST, FASTR1, FASTGRPAVGn) and SLOW readiout pattern
+            # (SLOW, SLOWR1) but there may not be CDPs corresponding exactly
+            # to these. Look only for CDPS matching the basic readout mode.
+            # TODO: Change this when CDPs become available for all modes.
+            #match_strings.append(readpatt)
+            if readpatt.startswith("FAST"):
```

Comment:
Does this mean that if someone has provided an explicit FASTR1 CDP file, the file matcher will still potentially or even preferentially (because of defaulting sorting) pick a "FAST" file instead, if present?

Preferably, we should try a match for the exact readpatt, and if that fails have a fall-back such as from FASTR1 to FAST.

Reply:
The change is an improvement in the sense that currently if you ask for a CDP file for data observed in FASTR1 or SLOWR1 mode you'll get a "CDP not found" error message. After this change, the software will match the CDPs already saved for "FAST" or "SLOW" mode. At the moment we only have CDPs for FAST or SLOW mode, so this is the appropriate thing to do. We could perhaps improve the implementation by logging a warning if you ask for FASTR1 but actually match FAST?

If in the future we have provided CDPs specifically for FASTR1 and SLOWR1 mode then it would make sense to go back to insisting on an exact match for READPATT - hence the "TODO". Implementing a searching system where it tries to find an exact match but falls back to FAST or SLOW if it can't would need to be implemented at a higher level (like the flat-field searching function within SCASim). But this is not straightforward to implement, so I think we should wait and see if it is needed.

Reply:
On second thoughts, I am wondering if we could implement the "try a match for the exact readpatt, and if that fails have a fall-back" at the file list filtering stage. The filter chooses the file with the most recent modification date, but it could also remove generic files from the list if specific files are available in the list. That would save having to implement a high-level try and search algorithm like we do for flat-fields.

Reply:
I guess the question is: which kind of software would use this "get_cdp" functionality, and would that software prefer to get an explicit "CDP not found" or would it prefer a fall-back? The latter may potentially cause confusion. I think adding the warning would make a lot of sense, but even then, not everyone will diligently read log files to see if warnings occurred. 

Perhaps in this case, it could be an explicit toggle for the "get_cdp" function that allows the user to choose whether a fall-back w.r.t. READPATT is allowed, defaulting to False to for now maintain existing behaviour. Any software that used "get_cdp" could then intentionally get updated to switch it to True if the author knows what they're doing and has confirmed that getting a FAST file when asking for FASTR1 is acceptable. What do you think?

I also agree that if it's not straightforward, then we should wait and see if/where it is needed.

Reply:
I have implemented the change I suggested in my second thoughts. The get_cdp function will now attempt to match the readpatt exactly. If an exact match is not found it issues a warning and falls back to either FAST or SLOW.

It wasn't easy to test this algorithm because there are no FASTR1 and SLOWR1 CDPs available on the ftp site, but I tested it by forcing get_cdp to look at the local CDP cache (where I had created some dummy FASTR1 and SLOWR1 versions).